### ISSUE TITLE
[FastPitch/Pytorch] Needing some help to quickstart my learning of analysing tensorboard graphs

### ISSUE THREAD

**Issue by cesm1980** [NVIDIA/DeepLearningExamples#843]
Ok, yesterday i just started training my first dataset for real (before this i was just using one temporary for tests and learning how to use things), i know its way too small (725 wavs/lines and ONLY 24 minutes in total), but from the little i already learning from interpretating tensorboard graphs, i THINK the model is learning, and i am consistently getting a reduction in loss, even tough at 400 steps it started to decrease more slowly, and now at step 2350 its on 0.11. From what i read in tutorials, i THINK the model isn't over fitting, but what i find strange is the validation dataset has a loss always stable at 2.5-2.6 and it doesn't seem to improve anymore, i have no idea if this is the so called overfitting or not, since its not increasing at all. (by the way it only has 71 lines, since i picked at random those from the main dataset, so the main one has 654).

I am using the default number of epochs on the train.sh script (1500 epochs), a batch size of 9 and 26 gradient-accumulation-steps (which made the epochs have 2 interactions, ie steps), i still dont have much idea of how exactly these things work so i choose those values because i wasnt able to use big batch size values due to having only 6gb vram card (gtx 1060 super), these were the fastest settings i was able to find (56s for doing a single epoch of two interances).

I just tested inference (almost finishing the 1500 epochs) and unfortunately i get half-giberish (actually sometimes i can understand what the voice says) and/or a lot echo/reverb on the voice, but to my amazement i DO recognize the voice being generated, so i think this is working even for such a SMALL dataset, it's just a matter of perhaps make some adjustments or simply doing more training, but first i would like if someone here is interested at looking at the tensorboards and help me learn how to spot problems or possible mistakes i am doing, because so far the online tutorials about tensorboard aren't DEFINITELY much "newbie-friendly" and most of them have different tag names on the logs than those used in the scripts of this repo for fastpitch, for example, in this repository i am not sure which graph is the attention that is always mentioned in other tutorials or issues from other repositories, the closest one that seems similar is the lrate one, also i only find the following ones :

aaa, frames, grad_max, grad_mean, grad_min, loss, lrate, mel_loss, took and zzz

Most of them it's common sense in what they are, and i already know that a decrease of loss is a good signal, and that a model is overfitting if the training loss stays the same or decreases and the validation loss increases, altough in my case the validation loss is stuck stable at 2.5 while the training loss is very slowly nearing zero.

Also i still have no idea what to interpret from grad_mean and grad_min. The frames and took graphs i think it's the time each step takes to be trained, regarding Lrate i know it means Learning rate, so far apparently it's the attention graph everyone says in other issues. It's odd that it was perfectly increasing diagonally just until 1k steps, and precisely after that started decreasing, although not as fast as it grew, not sure exactly what this means.

If anyone's interested in help just confirm before i post the graphs pics. By the way, @alancucki you said before you were willing to help, can you also help me with analysing the graphs? I think this isnt asking much, also i need to learn by myself how to do this as soon as possible, i just need a quickstart so that it gets easier to understand the tutorials about tensorboard, because most times they mention graphs that arent used in this training script, to analyse the quality/problems in training, which makes things harder to understand for a newbie.

**Comment by alancucki** [comment]
Regarding validation loss curves and overfitting: https://github.com/NVIDIA/DeepLearningExamples/issues/608 .

Unfortunately, with a generative model like this the best approach would be to listen to the samples and assess their quality on your own. In practive, loss and grad plots inform only that the training doesn't have critical issues.

For the default LJSpeech recipe, after 500/1500 epochs you should have already a decent model. Train on LJSpeech for some initial epochs and synthesize every 50/100 epochs - then you'll know what to expect early on from your own data.

For a small dataset, you might try to shrink the model, or add/augment the data. Again, you don't need to train for full 1500 epochs - if you're resourse-constrianed, compare early checkpoints.

Another thing to check would be the alignments (durations) for your dataset. You can analyze their distribution, look for outliers, and you can (roughly) compare the statistics to those of LJSpeech. Remember that Tacotron2 only works well for alignment extraction on the data, which it was trained on. You might need to train Tacotron2, fine-tune for your dataset, or look for other sources of alignments.

**Comment by cesm1980** [comment]
![loss](https://user-images.githubusercontent.com/42876695/109578083-a0973000-7aee-11eb-837d-b592da74d172.jpg)
![learning rate](https://user-images.githubusercontent.com/42876695/109578097-a4c34d80-7aee-11eb-838b-79fe29115489.jpg)
![mel loss](https://user-images.githubusercontent.com/42876695/109578107-a856d480-7aee-11eb-93f0-45a624bbcbff.jpg)

Oh thanks, but before i answer the things you said, here's the graphs for loss, learning rate and mel loss, i did 22900 steps at batch size 9 with gradient 26 (10 hours) and 3000 steps  with batch size 17 and gradient 1 (23 hours), the dataset in total is JUST 695 wavs, 618 on the training dataset and 77 on the validation dataset, unfortunately many of them are short, just 3-4 words, i think the total is only 27 minutes...

What you said still applys after seeing these graphs or you have anything extra to say ? Because i did so many tests (these were just two of them) and runs that i am starting to think that the problem is the dataset itself, since i tried these exact same settings with the ljspeech database and the loss decrease and inferences were MUCH better, nothing compared to this, so i am almost sure the problem is probably just the dataset itself...

After you comment on that i will reply the rest of the things to avoid too much writing at once.

**Comment by alancucki** [comment]
> 618 on the training dataset and 77 on the validation dataset

Like I said earlier, the validation dataset isn't much useful here. Once you know your hparams are roughly ok, I'd restart training on all 695 samples. It's more important to you to have as much data as possible.

Unfortunately, training loss is also poorly correlated with the perceived quality of synthesized speech. You can easily increase or decrease the loss just by changing dropout, yet with no gains in quality. With your data, you might either be overfitting, or failing to generalize; it's impossible to say with certainty. It depends no only on the volume, but also on the kind of speech, length distribution, quality of alignments, transcripts, etc.

I advise to structure the experiments, e.g., with the mentioned pairwise testing approach from the FastPitch paper. Check your hypotheses one by one and keep track of the results. Unfortunately, there are no shortcuts. It might turn out that the model is just too large for training from scratch on a small dataset.

**Comment by jnankp** [comment]
I am training my models from a remote server, how do I set up the tensorboard to monitor my training?

**Comment by alancucki** [comment]
Hi @jnankp,
you could run it directly on the remote server (with the `--bind_all` flag that runs on public network interfaces).

What I prefer to do instead is a periodical `rsync` of just tfevent files with the `--apend` flag, for instance every 10 minutes. That way you can compare multiple experiments in a single TensorBoard session, and have the convenience of running locally.
